pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,287 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslicemirroring
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller/endpointslicemirroring/metrics"
+	endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
+)
+
+// reconciler is responsible for transforming current EndpointSlice state into
+// desired state
+type reconciler struct {
+	client                clientset.Interface
+	maxEndpointsPerSubset int32
+	endpointSliceTracker  *endpointSliceTracker
+	metricsCache          *metrics.Cache
+	eventRecorder         record.EventRecorder
+}
+
+// reconcile takes an Endpoints resource and ensures that corresponding
+// EndpointSlices exist. It creates, updates, or deletes EndpointSlices to
+// ensure the desired set of addresses are represented by EndpointSlices.
+func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*discovery.EndpointSlice) error {
+	// Calculate desired state.
+	d := newDesiredCalc()
+
+	for _, subset := range endpoints.Subsets {
+		multiKey := d.initPorts(subset.Ports)
+
+		totalAddresses := 0
+		numInvalidAddresses := 0
+
+		for _, address := range subset.Addresses {
+			totalAddresses++
+			if totalAddresses > int(r.maxEndpointsPerSubset) {
+				break
+			}
+			if ok := d.addAddress(address, multiKey, true); !ok {
+				numInvalidAddresses++
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			}
+		}
+
+		for _, address := range subset.NotReadyAddresses {
+			totalAddresses++
+			if totalAddresses > int(r.maxEndpointsPerSubset) {
+				break
+			}
+			if ok := d.addAddress(address, multiKey, false); !ok {
+				numInvalidAddresses++
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			}
+		}
+
+		if numInvalidAddresses > 0 {
+			r.eventRecorder.Eventf(endpoints, corev1.EventTypeWarning, "InvalidIPAddress",
+				"Skipped %d invalid IP addresses when mirroring to EndpointSlices", numInvalidAddresses)
+		}
+	}
+
+	// Build data structures for existing state.
+	existingSlicesByKey := endpointSlicesByKey(existingSlices)
+
+	// Initialize data structures to track changes.
```

Comment:
comment is somewhat redundant

Reply:
Good call, pulled a comment from below up a bit instead.